src/uid2SecureSignal.ts

```diff
@@ -0,0 +1,61 @@
+export class Uid2SecureSignalProvider {
+  constructor() {
+    if (
+      typeof window.getUid2AdvertisingToken === "function" ||
+      // if the SDK has been initialized before this script loaded and enabled esp, the token will be
+      // send automatically as it might missed initComplete event 
+      this.isUID2SDKIntegrated()
+    ) {
+      this.registerSecureSignalProvider();
+    }
+  }
+
+  public registerSecureSignalProvider = () => {
+    const uid2Handler = this.retrieveAdvertisingTokenHandler();
+
+    if (!uid2Handler) {
+      console.warn("Please implement `getUid2AdvertisingToken`");
+      return;
+    }
+
+    if (uid2Handler()) {
+      window.googletag = window.googletag || {
+        cmd: [],
+      };
+      window.googletag.secureSignalProviders =
+        window.googletag.secureSignalProviders || [];
+      window.googletag.secureSignalProviders.push({
+        id: "uidapi.com",
+        collectorFunction: () => Promise.resolve(uid2Handler()),
```

Comment:
Is the goal to call `uid2Handler()` twice - once to make sure it returns something truthy, and also later when Secure Signals invokes the collector function?

Reply:
Yes, as when `secureSignalProviders.push` is called, the secureSignalProviders will immediately invoke the collector function. The first call could avoid this push get triggered when advertisting token is not ready and cache a invalid value such as undefined or null